include/execution.hpp

```diff
@@ -2171,7 +2171,222 @@ namespace std::execution {
+
+          if (__old != __completion_state) {
+            __self.__on_stop_.emplace(
+                get_stop_token(get_env(__self.__recvr_)),
+                __on_stop_requested{__shared_state->__stop_source_});
+          }
+
+          do {
+            if (__old == __completion_state) {
+              __self.__notify();
+              return;
+            }
+            __self.__next_ = static_cast<__operation_base*>(__old);
+          } while (!__head.compare_exchange_weak(
+              __old, static_cast<void *>(&__self),
+              memory_order_acq_rel,
+              memory_order_acquire));
+
+          if (__old == nullptr) {
+            start(__shared_state->__op_state2_);
+          }
+        } catch (...) {
+          execution::set_error((_Receiver&&) __self.__recvr_, current_exception());
+        }
+      };
+
+    template <class _SenderId>
+      class __sender {
+        using _Sender = __t<_SenderId>;
+        using __sh_state = __sh_state<_SenderId>;
+        template <class _Receiver>
+          using __operation = __operation<_SenderId, __x<remove_cvref_t<_Receiver>>>;
+
+        _Sender __sndr_;
+        shared_ptr<__sh_state> __shared_state_;
+
+      public:
+        template <__decays_to<__sender> _Self, receiver _Receiver>
+            requires receiver_of<_Receiver, completion_signatures_of_t<_Self>>
+          friend auto tag_invoke(connect_t, _Self&& __self, _Receiver&& __recvr)
+            noexcept(is_nothrow_constructible_v<decay_t<_Receiver>, _Receiver>)
+            -> __operation<_Receiver> {
+            return __operation<_Receiver>{(_Receiver &&) __recvr,
+                                          __self.__shared_state_};
+          }
+
+        template <__sender_queries::__sender_query _Tag, class... _As>
+            requires (!__is_instance_of<_Tag, get_completion_scheduler_t>) &&
+              __callable<_Tag, const _Sender&, _As...>
+          friend auto tag_invoke(_Tag __tag, const __sender& __self, _As&&... __as)
+            noexcept(__nothrow_callable<_Tag, const _Sender&, _As...>)
+            -> __call_result_if_t<__sender_queries::__sender_query<_Tag>, _Tag, const _Sender&, _As...> {
+            return ((_Tag&&) __tag)(__self.__sndr_, (_As&&) __as...);
+          }
+
+        template <class... _Tys>
+        using __set_value_t = completion_signatures<set_value_t(const decay_t<_Tys>&...)>;
+
+        template <class _Ty>
+        using __set_error_t = completion_signatures<set_error_t(const decay_t<_Ty>&)>;
+
+        template <__decays_to<__sender> _Self, class _Env>
+          friend auto tag_invoke(get_completion_signatures_t, _Self&&, _Env) ->
+            make_completion_signatures<
+              __member_t<_Self, _Sender>,
+              _Env,
+              completion_signatures<set_error_t(const exception_ptr&)>,
+              __set_value_t,
+              __set_error_t>;
+
+        explicit __sender(_Sender __sndr)
```

Comment:
Since the `split()` algorithm might heap allocate, do we need to consider adding support for custom allocators here?
Should we get the allocator from the sender?
And call `std::allocate_shared<__sh_state>(get_allocator(__sndr_), __sndr)`?

Or allow calling `split()` with an extra allocator argument?
And then call `std::allocate_shared<__sh_state>(__alloc, __sndr_)` instead.

Reply:
This is better dealt with as an issue.

Reply:
#463